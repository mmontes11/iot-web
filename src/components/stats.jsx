import React, { Fragment } from "react";

const Stats = () => <Fragment>{null}</Fragment>;

export default Stats;
