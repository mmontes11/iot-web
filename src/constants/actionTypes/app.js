export const TOGGLE_HAMBURGER_MENU = "TOGGLE_HAMBURGER_MENU";
export const TOGGLE_MAP_DIALOG = "TOGGLE_MAP_DIALOG";
