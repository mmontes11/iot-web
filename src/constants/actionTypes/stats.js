export const TYPE_SELECT = "TYPE_SELECT";
export const TYPE_UPDATED = "TYPE_UPDATED";