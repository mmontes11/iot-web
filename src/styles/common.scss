html {
  overflow: auto;
  background-color: $light;
}

.is-horizontal-center {
  display: flex;
  justify-content: center;
}

.is-selected {
  background-color: $selected;
}
