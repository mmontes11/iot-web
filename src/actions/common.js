import { RESET } from "constants/actionTypes/common";

export const reset = () => dispatch => {
  dispatch({ type: RESET });
};
