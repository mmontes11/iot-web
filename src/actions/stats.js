import { STATS_UPDATED, STATS_REQUEST, STATS_REQUEST_SUCCESS, STATS_REQUEST_ERROR } from "constants/actionTypes/stats";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import { normalizeStats } from "helpers/statsNormalizer";
import iotClient from "lib/iotClient";

const requestStats = (type, observation, onStart, onSuccess, onError) => {
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getStatsByType(observation)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getStatsByType(observation)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  }
};

export const getStats = () => (dispatch, getState) => {
  const {
    stats: {
      params: {
        type: { selectedItem: selectedType },
        observation: { selectedItem: selectedObservation },
      },
    },
  } = getState();
  if (!selectedType || !selectedObservation) {
    return;
  }
  dispatch({ type: STATS_REQUEST });
  requestStats(
    selectedType,
    selectedObservation,
    () => dispatch({ type: STATS_REQUEST }),
    res => {
      dispatch({ type: STATS_REQUEST_SUCCESS });
      let stats;
      const responseStats = res.body.stats;
      if (responseStats) {
        stats = responseStats.map(normalizeStats);
      }
      dispatch({ type: STATS_UPDATED, stats });
    },
    () => dispatch({ type: STATS_REQUEST_ERROR }),
  );
};
