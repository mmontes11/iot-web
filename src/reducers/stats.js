import { RESET } from "constants/actionTypes/common";
import paramsReducer from "reducers/params";

const initialState = {
  stats: null,
};

export default (state = initialState, action) => {
  const params = paramsReducer(state.params, action);
  switch (action.type) {
    case RESET:
      return { ...initialState, params };
    default:
      return { ...state, params };
  }
};
