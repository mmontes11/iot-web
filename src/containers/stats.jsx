import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { withResetOnUnmount } from "hocs/resetOnUnmount";
import { compose } from "recompose";
import PropTypes from "prop-types";
import StatsParamsPanel from "containers/statsParamsPanel";
import FiltersPanel from "components/filtersPanel";
import BarChart from "components/barChart";
import Loader from "components/loader";
import * as statsActions from "actions/stats";
import * as commonActions from "actions/common";
import * as reducerHelpers from "reducers";

const renderChart = (stats, isLoading) => {
  if (isLoading) {
    return <Loader />;
  } else if (stats) {
    return stats.map(statsItem => (
      <BarChart key={btoa(`${statsItem.title.type}-${statsItem.title.thing}`)} stats={statsItem} />
    ));
  }
  return null;
};

const Stats = ({ stats, isLoading, getStats, reset }) => (
  <div className="container is-fluid section">
    <div className="columns">
      <div className="column is-three-quarters">
        <StatsParamsPanel onParamsSelected={() => getStats()} onReset={() => reset()} />
        {renderChart(stats, isLoading)}
      </div>
      <div className="column is-one-quarter">
        <FiltersPanel />
      </div>
    </div>
  </div>
);

Stats.propTypes = {
  stats: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  isLoading: PropTypes.bool.isRequired,
  getStats: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
};

const withConnect = connect(
  state => ({
    stats: state.stats.items,
    isLoading: state.stats.isLoading && reducerHelpers.isLoading(state),
  }),
  { ...statsActions, ...commonActions },
);

export default compose(
  withConnect,
  withRouter,
  withResetOnUnmount,
)(Stats);
