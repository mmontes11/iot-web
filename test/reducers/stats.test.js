import deepFreeze from "deep-freeze";
import statsReducer from "reducers/stats";
import { RESET } from "constants/actionTypes/common";
import { STATS_REQUEST, STATS_REQUEST_SUCCESS, STATS_REQUEST_ERROR, STATS_UPDATED } from "constants/actionTypes/stats";
import { initialState } from "../constants/index";

const statsInitialState = initialState.stats;
deepFreeze(statsInitialState);

describe("reducers/stats", () => {
  it("has a default state", () => {
    expect(statsReducer(undefined, { type: "@init" })).toMatchSnapshot();
  });
  it("reduces nothing", () => {
    expect(statsReducer(statsInitialState, { type: "WHATEVER" })).toMatchSnapshot();
  });
  it("reduces STATS_REQUEST", () => {
    expect(statsReducer(statsInitialState, { type: STATS_REQUEST })).toMatchSnapshot();
  });
  it("reduces STATS_REQUEST_SUCCESS", () => {
    expect(statsReducer(statsInitialState, { type: STATS_REQUEST_SUCCESS })).toMatchSnapshot();
  });
  it("reduces STATS_REQUEST_ERROR", () => {
    expect(statsReducer(statsInitialState, { type: STATS_REQUEST_ERROR })).toMatchSnapshot();
  });
  it("reduces STATS_UPDATED", () => {
    expect(statsReducer(statsInitialState, { type: STATS_UPDATED, stats: [] })).toMatchSnapshot();
  });
  it("reduces RESET", () => {
    expect(statsReducer(statsInitialState, { type: RESET })).toMatchSnapshot();
  });
});
