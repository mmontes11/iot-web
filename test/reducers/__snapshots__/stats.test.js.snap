// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/stats has a default state 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "params": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
}
`;

exports[`reducers/stats reduces RESET 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "params": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
  "paramsInitialState": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
}
`;

exports[`reducers/stats reduces STATS_REQUEST 1`] = `
Object {
  "isLoading": true,
  "items": Array [],
  "params": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
}
`;

exports[`reducers/stats reduces STATS_REQUEST_ERROR 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "params": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
}
`;

exports[`reducers/stats reduces STATS_REQUEST_SUCCESS 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "params": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
}
`;

exports[`reducers/stats reduces STATS_UPDATED 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "params": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
}
`;

exports[`reducers/stats reduces nothing 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "params": Object {
    "observation": Object {
      "isActive": false,
      "isDisabled": true,
      "isLoading": false,
      "items": Array [],
      "selectedItem": null,
    },
    "reset": Object {
      "isDisabled": true,
    },
    "type": Object {
      "isActive": false,
      "items": Array [
        "event",
        "measurement",
      ],
      "selectedItem": null,
    },
  },
}
`;
